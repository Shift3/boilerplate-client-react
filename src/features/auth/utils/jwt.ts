import jwtDecode, { JwtPayload } from 'jwt-decode';

export const isMalformed = (token: string): boolean => {
  try {
    jwtDecode<JwtPayload>(token);
  } catch (e) {
    return true;
  }
  return false;
};

export const isExpired = (token: string): boolean => {
  const decoded = jwtDecode<JwtPayload>(token);
  const { exp } = decoded;
  // eslint-disable-next-line no-magic-numbers
  return !!exp && exp * 1000 < Date.now();
};