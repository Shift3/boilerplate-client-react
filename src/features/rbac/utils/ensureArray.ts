export const ensureArray = <T>(elem: T | T[]): T[] => (Array.isArray(elem) ? elem : [elem]);
