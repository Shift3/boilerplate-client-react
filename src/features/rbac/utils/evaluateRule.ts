import { User } from 'common/models';
import { Rule } from '../rules';

export const evaluateRule = (rule: Rule, user: User, data?: unknown): boolean => {
  if (typeof rule === 'function') {
    return rule(user, data);
  }

  return rule;
};
