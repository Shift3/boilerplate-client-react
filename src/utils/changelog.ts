export const hasOwnProperty = <X extends Record<string, unknown>, Y extends PropertyKey>(
  obj: X,
  prop: Y,
): obj is X & Record<Y, unknown> => {
  return Object.hasOwnProperty.call(obj, prop);
};

export const isObject = (obj: unknown): obj is Record<string, unknown> => {
  return obj !== null && typeof obj === 'object';
};

export const isStringOrNumber = (value: unknown): value is string | number => {
  return typeof value === 'string' || typeof value === 'number';
};

export const throwPropertyCannotBeFoundOrInvalidTypeError = (propertyName: string): Error => {
  return Error(`${propertyName} can't be found or change.${propertyName} is not a string or number`);
};

// For nestjs-changelog
export const changePropertyAccessor = (change: unknown, propertyName: string): string | number => {
  if (isObject(change) && !hasOwnProperty(change, 'createdAt')) {
    throw Error('Invalid change object');
  }

  if (isObject(change)) {
    if (propertyName === 'creationDate') {
      if (hasOwnProperty(change, 'createdAt') && isStringOrNumber(change.createdAt)) {
        return change.createdAt;
      }

      throw throwPropertyCannotBeFoundOrInvalidTypeError('createdAt');
    } else if (propertyName === 'editorName') {
      if (hasOwnProperty(change, 'whoDisplay') && isStringOrNumber(change.whoDisplay)) {
        return change.whoDisplay;
      }

      throw throwPropertyCannotBeFoundOrInvalidTypeError('whoDisplay');
    } else if (propertyName === 'id') {
      if (hasOwnProperty(change, 'id') && isStringOrNumber(change.id)) {
        return change.id;
      }

      throw throwPropertyCannotBeFoundOrInvalidTypeError('id');
    } else if (propertyName === 'itemType') {
      if (hasOwnProperty(change, 'itemType') && isStringOrNumber(change.itemType)) {
        return change.itemType;
      }

      throw throwPropertyCannotBeFoundOrInvalidTypeError('itemType');
    } else if (propertyName === 'changeAction') {
      if (hasOwnProperty(change, 'action') && isStringOrNumber(change.action)) {
        return change.action;
      }

      throw throwPropertyCannotBeFoundOrInvalidTypeError('action');
    } else {
      throw Error('Invalid property name given');
    }
  } else {
    throw Error('Given change parameter is not an object');
  }
};
// For nestjs-changelog
export const getChangeItemAction = (action: string): string => {
  switch (action) {
    case 'create':
      return 'create';
    case 'update':
      return 'update';
    case 'destroy':
      return 'delete';
    default:
      return action;
  }
};
