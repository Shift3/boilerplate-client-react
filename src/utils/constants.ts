import { environment } from '../environment/index';

interface IPattern {
  DIGIT_REGEX: RegExp;
  EMAIL_REGEX: RegExp;
  SYMBOL_REGEX: RegExp;
  US_PHONE_REGEX: RegExp;
  LOWERCASE_REGEX: RegExp;
  UPPERCASE_REGEX: RegExp;
}
interface IErrorMessages {
  FIELD_REQUIRED: string;
  PASSWORD_LENGTH: string;
  PASSWORD_LOWERCASE: string;
  PASSWORD_UPPERCASE: string;
  PASSWORD_SPECIAL_CHARACTER: string;
  PASSWORD_NUMBER: string;
  PASSWORD_MUST_MISMATCH: string;
  PASSWORD_MUST_MATCH: string;
  EMAIL_REQUIRED: string;
  INVALID_EMAIL: string;
  PASSWORD_REQUIRED: string;
  EMAIL_MATCH: string;
  FIRST_NAME_REQUIRED: string;
  LAST_NAME_REQUIRED: string;
  CURRENT_PASSWORD_REQUIRED: string;
  NEW_PASSWORD_REQUIRED: string;
  CONFIRM_PASSWORD_REQUIRED: string;
  ROLE_REQUIRED: string;
  SAME_EMAIL: string;
  VERIFICATION_CODE_REQUIRED: string;
  VERIFICATION_CODE_LENGTH_MISMATCH: string;
  VALID_PROFILE_PICTURE_FORMAT: string;
}

export interface IConstant {
  patterns: IPattern;
  errorMessages: IErrorMessages;
  version: string;
  creationYear: number;
  SUPPORTED_PROFILE_PICTURE_FORMATS: string[];
}

export const Constants: IConstant = {
  patterns: {
    DIGIT_REGEX: /[0-9]/,
    // eslint-disable-next-line no-useless-escape
    EMAIL_REGEX: /^[a-z0-9!#$%&'*+\/=?^_\`{|}~.-]+@[a-z0-9]([a-z0-9-])+(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$/i,
    SYMBOL_REGEX: /[-+_!@#$%^&*,.?]/,
    US_PHONE_REGEX: /(^\d{10}$)/,
    LOWERCASE_REGEX: /(?=.*[a-z])/,
    UPPERCASE_REGEX: /(?=.*[A-Z])/,
  },
  errorMessages: {
    FIELD_REQUIRED: 'Field is required.',
    PASSWORD_LENGTH: 'Password must be at least 8 characters.',
    PASSWORD_LOWERCASE: 'Password must contain at least one lowercase letter.',
    PASSWORD_UPPERCASE: 'Password must contain at least one uppercase letter.',
    PASSWORD_SPECIAL_CHARACTER: 'Password must contain at least one special character.',
    PASSWORD_NUMBER: 'Password must contain at least one number.',
    PASSWORD_MUST_MISMATCH: 'New password should not match current password.',
    PASSWORD_MUST_MATCH: 'Passwords must match.',
    EMAIL_MATCH: 'Email addresses must match.',
    FIRST_NAME_REQUIRED: 'First name is required.',
    LAST_NAME_REQUIRED: 'Last name is required.',
    EMAIL_REQUIRED: 'Email is required.',
    INVALID_EMAIL: 'Please enter a valid email address.',
    PASSWORD_REQUIRED: 'Password is required.',
    CURRENT_PASSWORD_REQUIRED: 'Current password is required.',
    NEW_PASSWORD_REQUIRED: 'New password is required.',
    CONFIRM_PASSWORD_REQUIRED: 'Confirm password is required.',
    ROLE_REQUIRED: 'Role is required.',
    SAME_EMAIL: 'Email must be different from current value.',
    VERIFICATION_CODE_REQUIRED: 'Verification code is required.',
    VERIFICATION_CODE_LENGTH_MISMATCH: 'Verification code must contain 6 digits.',
    VALID_PROFILE_PICTURE_FORMAT: 'Must select a supported image type (.jpeg, .jpg, or .png).',
  },
  version: environment.version,
  creationYear: 2021,
  SUPPORTED_PROFILE_PICTURE_FORMATS: ['image/jpg', 'image/jpeg', 'image/png'],
};
