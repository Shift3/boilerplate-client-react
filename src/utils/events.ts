export const wasMouseEventOutsideContainer = (container: HTMLElement, event: MouseEvent): boolean => {
  const boundingRect = container.getBoundingClientRect();

  return (
    event.clientX < boundingRect.left ||
    event.clientX > boundingRect.right ||
    event.clientY < boundingRect.top ||
    event.clientY > boundingRect.bottom
  );
};
