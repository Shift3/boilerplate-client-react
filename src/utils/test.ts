/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { JSXElementConstructor, ReactElement } from 'react';
import renderer from 'react-test-renderer';

const { getByTestId, getByRole, getByLabelText, queryAllByRole, queryByTestId, getAllByRole } = screen;

const { click, clear, type } = userEvent;

type expectToMatchSnapshotType = (component: ReactElement<any, string | JSXElementConstructor<any>>) => void;

export const expectToMatchSnapshot: expectToMatchSnapshotType = component =>
  expect(renderer.create(component).toJSON()).toMatchSnapshot();

export const expectInDocByTestId = (id: string) => expect(getByTestId(id)).toBeInTheDocument();

export const expectNotInDocByTestId = (id: string) => expect(queryByTestId(id)).not.toBeInTheDocument();

export const expectInDocByRole = (role: string) => expect(getByRole(role)).toBeInTheDocument();

export const expectInnerHTMLByTestId = (id: string, value: string) => expect(getByTestId(id).innerHTML).toEqual(value);

export const expectInnerHTMLByRole = (role: string, value: string) => expect(getByRole(role).innerHTML).toEqual(value);

export const clickByTestId = (id: string) => act(() => click(getByTestId(id)));

export const clickByTestIdAsync = (id: string) => act(async () => click(getByTestId(id)));

export const clickByRole = (role: string) => act(() => click(getByTestId(role)));

export const clickByRoleAsync = (role: string) => act(async () => click(getByRole(role)));

export const pathMatch = (path: string) => expect(window.location.pathname).toEqual(path);

export const clickNavigateByTestId = (id: string, path: string) => {
  clickByTestId(id);
  pathMatch(path);
};
export const clickNavigateByRole = (role: string, path: string) => {
  clickByRole(role);
  pathMatch(path);
};

export const expectTextContentByTestId = (id: string, text: string) =>
  expect(getByTestId(id)).toHaveTextContent(text.toString());

export const expectInDocByLabelText = (text: string) => expect(getByLabelText(text)).toBeInTheDocument();

export const expectLengthByRole = (role: string, length: number) => expect(queryAllByRole(role)).toHaveLength(length);

export const setValueByLabelText = (text: string, value: string) =>
  act(async () => {
    const element = getByLabelText(text);
    clear(element);
    type(element, value);
  });

export const formAlertMessageCheck = (message: string) => {
  const messages = getAllByRole('alert').reduce(
    (messages: string[], alert: HTMLElement) => [...messages, alert.innerHTML],
    [],
  );
  expect(messages.includes(message)).toBeTruthy();
};

export const expectMockFunctionCalled = (mockFunction: jest.Mock<any, any>) => expect(mockFunction).toHaveBeenCalled();

export const expectMockFunctionNotCalled = (mockFunction: jest.Mock<any, any>) =>
  expect(mockFunction).not.toHaveBeenCalled();

export const clickExpectByTestId = (clickId: string, expectId: string) => {
  clickByTestId(clickId);
  expectInDocByTestId(expectId);
};
