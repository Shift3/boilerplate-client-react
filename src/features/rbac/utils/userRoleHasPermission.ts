import { User } from 'common/models';
import { Permission, RbacRules } from '../rules';
import { evaluateRule } from './evaluateRule';

export const userRoleHasPermission = (
  rules: RbacRules,
  permission: Permission,
  user: User,
  data?: unknown,
): boolean => {
  const permissionMap = rules[user.role];

  if (!permissionMap) {
    return false;
  }

  const rule = permissionMap[permission];

  if (!rule) {
    return false;
  }

  return evaluateRule(rule, user, data);
};
