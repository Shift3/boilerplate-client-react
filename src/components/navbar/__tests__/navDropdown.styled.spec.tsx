import { expectToMatchSnapshot } from '../../../utils/test';
// import renderer from 'react-test-renderer';
import {
  DropdownButton,
  Dropdown,
  NavDropdownItem,
  DropdownContainerLeft,
  ProfilePhoto,
  DropdownContainerRight,
  VerticalLine,
  // NavIcon,
  // DirectoryLink,
} from '../styled/navDropdown.styled';

import 'jest-styled-components';

describe('<NavDropdown /> styled component snapshot tests', () => {
  it ('Should match the stored <DropdownButton /> snapshot', () =>
    expectToMatchSnapshot(<DropdownButton/>));

  it ('Should match the stored <Dropdown /> snapshot', () =>
    expectToMatchSnapshot(<Dropdown/>));

  it ('Should match the stored <NavDropdownItem /> snapshot', () =>
    expectToMatchSnapshot(<NavDropdownItem/>));

  it ('Should match the stored <DropdownContainerLeft /> snapshot', () =>
    expectToMatchSnapshot(<DropdownContainerLeft/>));

  it ('Should match the stored <ProfilePhoto /> snapshot', () =>
    expectToMatchSnapshot(<ProfilePhoto/>));

  it ('Should match the stored <DropdownContainerRight /> snapshot', () =>
    expectToMatchSnapshot(<DropdownContainerRight/>));

  it ('Should match the stored <VerticalLine /> snapshot', () =>
    expectToMatchSnapshot(<VerticalLine/>));

  // @TODO - find solution for snapshot testing with components
  // that do not fit the type definition for currently used methodology

  // it ('Should match the stored <NavIcon /> snapshot', () =>
  //   expect(renderer.create(<NavIcon />).toJSON()).toMatchSnapshot());

  // it ('Should match the stored <DirectoryLink /> snapshot', () =>
  //   expect(renderer.create(<DirectoryLink />).toJSON()).toMatchSnapshot());
});