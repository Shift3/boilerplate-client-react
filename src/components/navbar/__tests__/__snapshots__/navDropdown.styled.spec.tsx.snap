// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <Dropdown /> snapshot 1`] = `
.c0 a {
  color: rgba(0,0,0,.5);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 a:hover {
  color: rgba(0,0,0,.7);
}

.c0 .dropdown-menu {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .dropdown-menu a {
  color: #000;
}

.c0 .dropdown-menu .dropdown-item > a:hover {
  background-color: rgba(0,0,0,0);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0 nav-item dropdown"
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle nav-link"
    href="#"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  />
</div>
`;

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <DropdownButton /> snapshot 1`] = `
.c0 {
  background-color: rgba(0,0,0,0);
  border-color: rgba(0,0,0,0);
}

.c0:hover,
.c0:focus,
.c0:active {
  background-color: rgba(0,0,0,0) !important;
}

.c0:active,
.c0:focus {
  border-color: #175f6e;
}

.c0:focus,
.c0:hover {
  border-color: rgba(0,0,0,0);
}

<button
  className="c0 btn btn-primary"
  disabled={false}
  type="button"
/>
`;

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <DropdownContainerLeft /> snapshot 1`] = `
.c0 {
  padding: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
}

<div
  className="c0"
/>
`;

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <DropdownContainerRight /> snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  className="c0"
/>
`;

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <NavDropdownItem /> snapshot 1`] = `
.c0:active,
.c0:focus {
  background-color: rgba(233,236,239,0);
}

.c0.dropdown-item.active,
.c0 .dropdown-item:active {
  background-color: rgba(0,0,0,0) !important;
}

<a
  className="c0 dropdown-item"
  href="#"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
/>
`;

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <ProfilePhoto /> snapshot 1`] = `
.c0 {
  border-radius: 50%;
  max-height: 50px;
  max-width: 50px;
  width: auto;
  height: auto;
}

<img
  className="c0"
/>
`;

exports[`<NavDropdown /> styled component snapshot tests Should match the stored <VerticalLine /> snapshot 1`] = `
.c0 {
  margin: 0 20px;
  height: 125px;
  border-left: 1px solid #d3d3d3;
  display: inline-block;
}

<span
  className="c0"
/>
`;
